### DISCUSSION TITLE
Change the state of a field before submitting the form.

### DISCUSSION THREAD

**Discussion by pabbec** [filamentphp/filament#9704] [Help] [answered]
I'm using Filament 3, PHP 8.1 and Laravel 10.x. 

I was able to add two buttons at form bottom overwriten getFormActions method (following some examples in this forum). But I need to modify the value (state) of a field depending on the button the user clicks. I explain my use case:

I have two buttons in my form:

- Save to draft
- Save and Publish

In both cases, I want to submit the form to save the data. But, if the user clicks the "Save to draft" button, I need change the status field to the value "draft". And if the user clicks the "Save and publish" button, I need to set the status field to the value "published".

In my PostResource/Pages/CreatePost.php I have:

```
     protected function getFormActions(): array
     {
         return [
             $this->getCreateInDraftFormAction(),
             $this->getCreateAndConfirmFormAction(),
             $this->getCancelFormAction(),
         ];
     }

     protected function getCreateInDraftFormAction(): Action
     {
         return Action::make('createindraft')
         ->label('Save to Draft')
         ->submit('createindraft');
     }

     protected function getCreateAndConfirmFormAction(): Action
     {
         return Action::make('createandconfirm')
         ->label('Save and Confirm')
         ->submit('createandconfirm');
     }
```

I can't find a way to access that field in my form and set the value I want.

Can you help me with this?

**Comment by sinnbeck** [comment]
Maybe see if this is available for create as well :)

https://filamentphp.com/docs/3.x/panels/resources/editing-records#customizing-data-before-saving

**Comment by pabbec** [comment]
Thanks@sinnbeck. Yes, it is an option. But I couldn't find a way to detect which of the two send buttons (Save to Draft or Save and Publish) the user clicked to be able to change the state of that field.

If there is a way to detect that, I would appreciate it if you could explain it to me.

**Reply by sinnbeck** [reply]
Ah yes I see.

Took a bit to figure out but think this should do it
```php
    protected function getCreateInDraftFormAction(): Action
    {
        return Action::make('createindraft')
            ->label('Save to Draft')
            ->action(function () {
                $this->fill(['draft' => true]);
                $this->create(); //or $this->save(); for edit pages
            });
    }

**Comment by pabbec** [comment]
I have tried but it throws an exception:

BadMethodCallException: Method App\Filament\Resources\PostResource\Pages\CreatePost::save does not exist.

```
     protected function getFormActions(): array
     {
         return [
             $this->getCreateInDraftFormAction(),
             $this->getCreateAndConfirmFormAction(),
             $this->getCancelFormAction(),
         ];
     }

protected function getCreateAndConfirmFormAction(): Action
    {
        return Action::make('createandconfirm')
            ->label('Save and confirm')
            ->action(function () {
                $this->fill(['status' => 'confirmed']);
              return $this->save();
            });
    }

protected function getCreateInDraftFormAction(): Action
    {
        return Action::make('createandconfirm')
            ->label('Save in draft')
            ->action(function () {
                $this->fill(['status' => 'draft']);
              return $this->save();
        });
    }

```
I feel like there is something missing from the documentation, :)

**Reply by sinnbeck** [reply]
I have updated my answer. I tested on an edit page. Its ->create()

**Reply by pabbec** [reply, marked as answer]
Thank you so much!

However $this->fill(['status'] => 'published'); The state does not change (I don't know why), but while debugging, I see that the "data" property exists, and what I did was access it directly:

```
protected function getCreateAndConfirmFormAction(): Action
    {
        return Action::make('createandconfirm')
            ->label('Save and confirm')
            ->action(function () {
                //$this->fill(['status' => 'confirmed']);
              $this->data['status'] = 'published';
              return $this->create();
            });
    }
```
I'm not sure if this is valid or not, but it works :)

Again, thanks for your time.

**Reply by sinnbeck** [reply]
Ah ok. Didnt have a draft column so just assumed :D But yeah that should be completely fine as well :)

**Comment by rasul-chopurov** [comment]
не работает :/
